### PR TITLE
Refactor settings of p:f:t

### PR DESCRIPTION
**Pull request by peterrum**
This PR has a look at the settings of `p:f:t` as motivated by @bangerth in PR #9007 . Currently there are two settings:
```cpp
    enum Settings
    {
      default_setting = 0x0,
      construct_multigrid_hierarchy = 0x1
    };
```
Enabling `Settings::construct_multigrid_hierarchy` leads currently directly in enabling `Triangulation<dim, spacedim>::limit_level_difference_at_vertices`, leading to the issue that `p:f:t:` might have more (artificial) cells than the base triangulation it was constructed from (see issue #9353). Which is quite confusing for users.

I therefore would like to prose that `TriangulationDescription::Description` (see #9195) also contains the `MeshSmoothing` info (and in particular the `Setting` field is removed). @luca-heltai what do you think?

~Also I would like to eliminate the  input argument `construct_multilevel_hierarchy` from the function `TriangulationDescription::Utilities::create_description_from_triangulation()`. When this argument is set to `false`,~ the input triangulation is on the fly flattened and for this flattened triangulation the `Description` is created. This option is only applicable for triangulation created by global refinements and is not general. Since the flattening may lead to the generation new `CellIds` this setting also might lead to confusion and should be therefore removed. If we remove this flags, we also need to refactor less code later on (see issue #9390) ;)

~This PR is building on top of the not yet merged PR #9195 and will be rebased as soon as possible.~

FYI @agrayver

### CODE REVIEW

**Comment by peterrum** [comment]
The relevant test is *tests/fullydistributed_grids/copy_serial_tria_09.cc*.

**Comment by bangerth** [comment]
Please rebase.

**Comment by peterrum** [comment]
@bangerth Done!

File: include/deal.II/grid/tria_description.h [hunk 1]
```diff
@@ -398,6 +398,8 @@ namespace TriangulationDescription
      * Properties to be use in the construction of the triangulation.
      */
     Settings settings;
+
+    typename Triangulation<dim, spacedim>::MeshSmoothing smoothing;
```
Comment (bangerth) [inline include/deal.II/grid/tria_description.h]: Can you add a comment?

**Comment by bangerth** [comment]
Is this really all you wanted to do, or did anything get lost in the rebase?

File: tests/fullydistributed_grids/copy_serial_tria_09.cc [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+  for (auto &cell : tria_pft.active_cell_iterators())
+    {
+      CellId id        = cell->id();
+      auto   cell_base = id.to_cell(basetria);
+      for (unsigned int d = 0; d < dim; d++)
+        Assert(std::abs(cell->center()[d] - cell_base->center()[d]) < 1e-9,
+               ExcMessage("Cells do not match"));
+    }
+}
```
Comment (bangerth) [inline tests/fullydistributed_grids/copy_serial_tria_09.cc]: We typically output "OK" at the end of the test function for tests that check everything via assertions, rather than by producing some output.

File: include/deal.II/grid/tria_description.h (hunk 1 above)
Comment (peterrum) [reply include/deal.II/grid/tria_description.h]: @bangerth I have not finished this PR, yet. See the description of this PR. Also, I would be interested if you and @luca-heltai agree on having in `TriangulationDescription::Description` the information regarding to mesh smoothing. `TriangulationDescription::Description` is in some form used by all triangulations now...

Comment (bangerth) [reply include/deal.II/grid/tria_description.h]: If you include the settings into the structure then the smoothing flags make sense as well.

Comment (peterrum) [reply include/deal.II/grid/tria_description.h]: `Settings` is actually what will be removed from the struct and moved to the constructor of `p:f:t` as you suggested in an other PR.

Comment (bangerth) [reply include/deal.II/grid/tria_description.h]: OK, and I still agree with that. Then I would also move the flags along with the settings.

**Comment by bangerth** [comment]
So if you move the settings, then the flags should go as there as well and nothing of the current state of the patch is left...

**Comment by bangerth** [comment]
Ping?

**Comment by peterrum** [comment]
@bangerth Give me a couple of days. I am too occupied with a couple of other projects.

**Comment by peterrum** [comment]
I will work on this PR so that we can finalize the settings of `p:f:t` before the 9.2 release.

**Comment by peterrum** [comment]
@bangerth Sorry for the delay of this PR (there have been always tasks with higher priority). It is now ready to be reviewed. Thanks!

Let me review the purpose and reason of this PR. 

@agrayver noticed that p:f:t, when initialized by a serial mesh, was adding additional cells. This was related to mismatching mesh smoothing settings (`none` vs. `limit_level_difference_at_vertices`). This is not a problem as such (since it does not lead to wrong results), however, is confusing for users when traversing over the cells in the source triangulation and in p:f:t .

This PR improves this circumstance by copying the mesh smoothness setting from the source triangulation and by only amending this setting by `limit_level_difference_at_vertices` if `Settings::construct_multigrid_hierarchy`is requested (similarly as in the case of p:d:t).

Furthermore, this PR introduces a second change. While not setting `Settings::construct_multigrid_hierarchy` used to mean that the triangulation is flattened, this is not done any more. (I don't see the purpose of this flattening any more since there is function in `GridTools` exactly for this purpose.)

In conclusion: Independent of if `Settings::construct_multigrid_hierarchy` is set or not, p:f:t is set up now with all levels. Setting `Settings::construct_multigrid_hierarchy` influences only the mesh smoothing setting.

**Comment by peterrum** [comment]
By the way this is related to @tjhei's comment https://github.com/dealii/dealii/pull/8622#discussion_r316937748 which is did not get at that time.

**Comment by agrayver** [comment]
@peterrum thanks for addressing this. 

I have a more general question. What dictates your decision to enforce `limit_level_difference_at_vertices` other than the compatibility with p::d::f? Is there any fundamental reason to have this condition satisfied when requesting `Settings::construct_multigrid_hierarchy`?

**Comment by peterrum** [comment]
> What dictates your decision to enforce limit_level_difference_at_vertices other than the compatibility with p::d::f? Is there any fundamental reason to have this condition satisfied when requesting Settings::construct_multigrid_hierarchy?

I would have thought it is requited to be able to run multigrid on distributed meshes (so that all ghost cells are available to all levels). But maybe my assumption is not correct and there is an other 
reason (like p4test, ...). 

@tjhei could you help me and shortly clarify why `p:d:t` sets `limit_level_difference_at_vertices` when `construct_multigrid_hierarchy` is enabled. Thanks!

**Comment by masterleinad** [comment]
> @tjhei could you help me and shortly clarify why `p:d:t` sets `limit_level_difference_at_vertices` when `construct_multigrid_hierarchy` is enabled. Thanks!

Have a look at https://github.com/dealii/dealii/issues/310.

**Comment by tjhei** [comment]
> Have a look at #310.

Yes, and this has nothing to do with p4est if I remember correctly.

**Comment by masterleinad** [comment]
Yes, we need to make sure that the multigrid transfer only needs to deal with one level at a time. For continuous elements, this implies also that the level of cells meeting at a vertex can not differ by more than one. We make sure that this is the case for all triangulations when we construct the multigrid hierarchy.

**Comment by peterrum** [comment]
@masterleinad @tjhei Thanks for the useful links! This is what I was also thinking. 

@agrayver did the comments answer your question?

**Comment by agrayver** [comment]
> @agrayver did the comments answer your question?

Yes, thanks. Now, the logic of p::f::t construction fully makes sense to me.

**Comment by peterrum** [comment]
Any opinions on this PR?

File: tests/fullydistributed_grids/copy_distributed_tria_01.with_p4est=true.mpirun=5.output [hunk 3]
```diff
@@ -1,77 +1,77 @@
...
+DEAL:4:2d::n_cells:                   1373
+DEAL:4:2d::n_active_cells:            1030
 DEAL:4:2d::
 DEAL:4:2d::n_dofs:                             16641
 DEAL:4:2d::n_locally_owned_dofs:               3280
 DEAL:4:2d::
-DEAL:4:3d::n_levels:                  1
-DEAL:4:3d::n_cells:                   1108
-DEAL:4:3d::n_active_cells:            1108
+DEAL:4:3d::n_levels:                  5
```
Comment (tjhei) [inline tests/fullydistributed_grids/copy_distributed_tria_01.with_p4est=true.mpirun=5.output:72]: I am not sure I understand the changes here. Did you check that this makes sense?

File: tests/fullydistributed_grids/copy_serial_tria_09.cc [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// Create a serial triangulations with different mesh-smoothing flags levels
```
Comment (tjhei) [inline tests/fullydistributed_grids/copy_serial_tria_09.cc:17]:
```suggestion
// Create a serial triangulations with different mesh-smoothing flags 
```

File: source/grid/tria_description.cc [hunk 5]
```diff
@@ -178,134 +178,21 @@ namespace TriangulationDescription
...
-            }
-
-          // 4) correct vertices of cells (make them local)
-          for (auto &cell : construction_data.coarse_cells)
-            for (const unsigned int v : GeometryInfo<dim>::vertex_indices())
-              cell.vertices[v] = vertices_locally_relevant[cell.vertices[v]];
-        }
-      else
+      // process cells and vertices
+      if (true /*TODO: prevent change of indentation*/)
```
[the remark this answers is not in the document]
Comment (peterrum) [reply source/grid/tria_description.cc:194]: I left this if statement in so that the code there is not reformatted and shown as modified code in the review. I will remove the if statement now!

File: tests/fullydistributed_grids/copy_distributed_tria_01.with_p4est=true.mpirun=5.output (hunk 3 above)
Comment (peterrum) [reply tests/fullydistributed_grids/copy_distributed_tria_01.with_p4est=true.mpirun=5.output:72]: Yes these changes make sense. There used to be the option to flatten the triangulation. This PR removed this option and keeps the levels from the original mesh. In the diff above, you can see the result: the number of levels has increased from 1 to 5.

**Comment by peterrum** [comment]
@tjhei and @luca-heltai Thanks for the review! I have made the changes!
